tests/harness/utils.py

```diff
@@ -0,0 +1,349 @@
+    yield client
+
+    # remove all traces, and change working directory back to base path
+    os.chdir(orig_cwd)
+    client.activate_root(orig_root)
+    cleanup_folder(str(tmp_path))
+
+
+@contextmanager
+def clean_project_session(
+    tmp_path_factory: pytest.TempPathFactory,
+    clean_repo: bool = False,
+) -> Generator[Client, None, None]:
+    """Context manager to create, activate and use a separate ZenML project.
+
+    Args:
+        tmp_path_factory: A pytest fixture that provides a temporary directory.
+        clean_repo: Whether to create and use a clean repository for the
+        project.
+
+    Yields:
+        A ZenML client configured to use the project.
+    """
+    from zenml.utils.string_utils import random_str
+
+    client = Client()
+    original_project = client.active_project.id
+
+    project_name = f"pytest_{random_str(8)}"
+    client.create_project(name=project_name, description="pytest test project")
+
+    if clean_repo:
+        with clean_repo_session(tmp_path_factory) as repo_client:
+            repo_client.set_active_project(project_name)
+
+            logging.info(f"Tests are running in project: '{project_name}'")
+            yield repo_client
+    else:
+        client.set_active_project(project_name)
+
+        logging.info(f"Tests are running in project: '{project_name}'")
+        yield client
+
+    # change the active project back to what it was
+    client.set_active_project(original_project)
+    client.delete_project(project_name)
+
+
+@contextmanager
+def clean_default_client_session(
+    tmp_path_factory: pytest.TempPathFactory,
+) -> Generator[Client, None, None]:
+    """Context manager to initialize and use a clean local default ZenML client
+    with its own global configuration and local database.
+
+    Args:
+        tmp_path_factory: A pytest fixture that provides a temporary directory.
+
+    Yields:
+        A clean ZenML client.
+    """
+
+    # save the current global configuration and client singleton instances
+    # to restore them later, then reset them
+    orig_cwd = os.getcwd()
+    original_config = GlobalConfiguration.get_instance()
+    original_client = Client.get_instance()
+    orig_config_path = os.getenv("ZENML_CONFIG_PATH")
+
+    GlobalConfiguration._reset_instance()
+    Client._reset_instance()
+
+    # change the working directory to a fresh temp path
+    tmp_path = tmp_path_factory.mktemp("pytest-clean-client")
+    os.chdir(tmp_path)
+
+    os.environ["ZENML_CONFIG_PATH"] = str(tmp_path / "zenml")
+    os.environ["ZENML_ANALYTICS_OPT_IN"] = "false"
+
+    # initialize the global config client and store at the new path
+    gc = GlobalConfiguration()
+    gc.analytics_opt_in = False
+    client = Client()
+    _ = client.zen_store
+
+    logging.info(f"Tests are running in clean environment: {tmp_path}")
+
+    yield client
+
+    # restore the global configuration path
+    if orig_config_path:
+        os.environ["ZENML_CONFIG_PATH"] = orig_config_path
```

Comment:
This could probably also use the defined constant from `zenml.constants`?

Reply:
done